clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/UpsertGoogleScalingPolicyAtomicOperation.groovy

```diff
@@ -0,0 +1,124 @@
+  */
+
+  @Override
+  Void operate(List priorOutputs) {
+    def serverGroupName = description.serverGroupName
+    task.updateStatus BASE_PHASE, "Initializing upsert of scaling policy for $serverGroupName..."
+
+    def credentials = description.credentials
+    def project = credentials.project
+    def compute = credentials.compute
+    def accountName = description.accountName
+    def region = description.region
+    def serverGroup = GCEUtil.queryServerGroup(googleClusterProvider, accountName, region, serverGroupName)
+    def isRegional = serverGroup.regional
+    def zone = serverGroup.zone
+    def location = isRegional ? region : zone
+    def autoscaler = buildAutoscaler(credentials, project, location, serverGroupName, isRegional)
+
+    task.updateStatus BASE_PHASE, "Attempting to retrieve existing autoscaler for $serverGroupName..."
+
+    def scalingPolicies
+    if (isRegional) {
+      scalingPolicies = compute.regionAutoscalers().list(project, region).execute().getItems()
+    } else {
+      scalingPolicies = compute.autoscalers().list(project, zone).execute().getItems()
+    }
+
+    def scalingPolicyExistsForMIG = scalingPolicies.any { it.name == serverGroupName }
+
+    def operation
+    if (scalingPolicyExistsForMIG) {
+      task.updateStatus BASE_PHASE, "Updating existing autoscaler $serverGroupName..."
+
+      if (isRegional) {
+        operation = compute.regionAutoscalers().update(project, region, autoscaler).execute()
+      } else {
+        operation = compute.autoscalers().update(project, zone, autoscaler).execute()
+      }
+    } else {
+      task.updateStatus BASE_PHASE, "Creating new autoscaler $serverGroupName..."
+
+      if (isRegional) {
+        operation = compute.regionAutoscalers().insert(project, region, autoscaler).execute()
+      } else {
+        operation = compute.autoscalers().insert(project, zone, autoscaler).execute()
+      }
+    }
+
+    def operationName = operation.getName()
+    if (isRegional) {
+      googleOperationPoller.waitForRegionalOperation(compute, project, region, operationName, null, task,
+        "regional autoscaler $serverGroupName", BASE_PHASE)
+    } else {
+      googleOperationPoller.waitForZonalOperation(compute, project, zone, operationName, null, task,
+        "zonal autoscaler $serverGroupName", BASE_PHASE)
+    }
+
+    null
+  }
+
+  Autoscaler buildAutoscaler(GoogleNamedAccountCredentials credentials, String project, String location, String serverGroupName, Boolean isRegional) {
+    def managedInstanceGroup
```

Comment:
Seems like the only reason to make the live calls to GCE in this method are to retrieve the `zone` and `selfLink`.
You already have the server group and it exposes `selfLink`, so we can just use that.
It also exposes `zone`, but it will be null for regional server groups. I think the original `buildAutoscaler`was wrong to expect `zone` in all cases, as it should be `region` for regional server groups. This can be fixed as part of this pr as well.
So this method should retrieve the attributes it needs from the `GoogleServerGroup` object and pass `zone` or `region` to an updated `GCEUtil.buildAutoscaler` method.